### PR TITLE
Refactor stbt batch run

### PR DESCRIPTION
**Pull request by wmanley**
Continuing the saga of pythonising `stbt batch run`.  This should be faster as there is less spawning processes.  I'd like to be able to replace all the little files with a single JSON file and this refactor.

### CODE REVIEW

File: stbt-batch.d/run-one [hunk 1]
```diff
@@ -110,22 +109,9 @@ main() {
     "$runner"/report --html-only . >/dev/null
   fi
 
-  cd ..
```
[the remark this answers is not in the document]
Comment (wmanley) [reply stbt-batch.d/run-one]: Actually this is quite deliberate.  We don't want cwd to be inside the directory when we send the message that the directory is complete.

File: stbt-batch.d/run-one [hunk 2]
```diff
@@ -110,22 +109,9 @@ main() {
...
-send_state_change() {
-  if [ -z "$STBT_TRACING_SOCKET" ]; then
-    return
-  fi
-
-  date='"'$(date --iso-8601=ns)'"'
-  echo '{"state_change": {"changes": {"'"$1"'": '"$2"'}, "time": '"$date"'}}' \
-    | socat STDIN "UNIX-CONNECT:$STBT_TRACING_SOCKET"
-  true
-}
```
[the remark this answers is not in the document]
Comment (wmanley) [reply stbt-batch.d/run-one]: Done

File: stbt-batch.d/run.py [hunk 3]
```diff
@@ -22,8 +22,6 @@
 
 
 def main(argv):
-    runner = os.path.dirname(os.path.abspath(__file__))
```
[the remark this answers is not in the document]
Comment (wmanley) [reply stbt-batch.d/run.py]: Done

File: stbt-batch.d/run.py [hunk 4]
```diff
@@ -194,16 +192,29 @@ def run_one(test, args, tag, cwd)
     """
     Invoke the run-one shell-script with the appropriate arguments.
     """
+    test_file = os.path.abspath(test[0])
+    test_args = list(test[1:])
```
[the remark this answers is not in the document]
Comment (wmanley) [reply stbt-batch.d/run.py]: Done, but I'm not sure that it's an improvement.

File: stbt-batch.d/run.py [hunk 5]
```diff
@@ -152,6 +150,119 @@ def on_term(_signo, _frame):
...
+        cmd += ['-vv']
+    else:
+        cmd += ['-v']
+    cmd += [test_file, '--'] + list(test_args)
+
+    subenv = dict(os.environ)
+    subenv['do_html_report'] = "true" if batch_args.do_html_report else "false"
+    subenv['stbt_root'] = _find_file('..')
+    subenv['test_displayname'] = " ".join(
+        (os.path.basename(test_name),) + test_args)
```
Comment (drothlis) [inline stbt-batch.d/run.py]: This is a regression in behaviour (from master, not from the previous commit). We should display the test file as given on the command line -- not the basename, and not the abspath either. I'll push a fixup.

File: stbt-batch.d/run.py [hunk 6]
```diff
@@ -152,6 +150,119 @@ def on_term(_signo, _frame):
...
+    def write_file(name, data):
+        with open(os.path.join(rundir, name), 'w') as f:
+            f.write(data)
+
+    if git_info:
+        write_file("git-commit", git_info.commit)
+        write_file("git-commit-sha", git_info.commit_sha)
+        write_file("test-name", os.path.relpath(test_name, git_info.git_dir))
+    else:
+        write_file("test-name", test_name)
```
Comment (drothlis) [inline stbt-batch.d/run.py]: Also a regression here: If there's no `.git` then we used to write the absolute path to the file.

**Pull request merged**